kernel: system workqueue stack size as maximum

JordanYates opened the pull request:
As a co-operative workqueue, the system workqueue stack size should be
set as the maximum value requested by any of the users. Users request a
minimum size for the workqueue stack by defining a Kconfig symbol of the
form `SYSTEM_WORKQUEUE_MIN_STACK_SIZE_`. The build system then sets the
stack size to the largest of the requested values.

In order to maintain compatability with existing applications, the value
of `SYSTEM_WORKQUEUE_STACK_SIZE` is copied into a kconfig symbol that
matches the above format.

This behaviour can be disabled if required through
`SYSTEM_WORKQUEUE_STACK_SIZE_EXPLICIT`.

Implements #88140

JordanYates committed:
kernel: system workqueue stack size as maximum

As a co-operative workqueue, the system workqueue stack size should be
set as the maximum value requested by any of the users. Users request a
minimum size for the workqueue stack by defining a Kconfig symbol of the
form `SYSTEM_WORKQUEUE_MIN_STACK_SIZE_`. The build system then sets the
stack size to the largest of the requested values.

In order to maintain compatability with existing applications, the value
of `SYSTEM_WORKQUEUE_STACK_SIZE` is copied into a kconfig symbol that
matches the above format.

This behaviour can be disabled if required through
`SYSTEM_WORKQUEUE_STACK_SIZE_EXPLICIT`.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
kernel: move workqueue stack size defaults

Move the defaults for `SYSTEM_WORKQUEUE_STACK_SIZE` to the appropriate
subsystem kconfig files.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Obviously this requires docs and migration guide notes, looking for feedback on the implementation first

yperess reviewed (changes requested):
This feels like the kind of "magic" naming that people complain about a lot and makes Zephyr (in my experience) hard to use. Why can't we just add to these features `depends on SYSTEM_WORKQUEUE_STACK_SIZE >= 2560` for the WPA_SUPPLICANT and `depends on SYSTEM_WORKQUEUE_STACK_SIZE >= 4096` for the COVERAGE_GCOV?

JordanYates commented:
> This feels like the kind of "magic" naming that people complain about a lot and makes Zephyr (in my experience) hard to use. Why can't we just add to these features `depends on SYSTEM_WORKQUEUE_STACK_SIZE >= 2560` for the WPA_SUPPLICANT and `depends on SYSTEM_WORKQUEUE_STACK_SIZE >= 4096` for the COVERAGE_GCOV?

There are a few reasons.

Firstly, `depends on SYSTEM_WORKQUEUE_STACK_SIZE >= 4096` for `COVERAGE_GCOV` would simply result in the majority of coverage test builds failing, because most of them are not going to satisfy the requirement.

Secondly, it makes choosing a value a giant pain for applications that are hardware agnostic. Pretend you want to add `k_work_submit()` to `zephyr/samples/hello_world`, what value should you assign to `CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE` in `prj.conf`, given the below requirements?
https://github.com/zephyrproject-rtos/zephyr/blob/9c5abb35b5ffb0326e89e2de0504bad8b9253d5d/arch/arc/Kconfig#L398-L399
https://github.com/zephyrproject-rtos/zephyr/blob/9c5abb35b5ffb0326e89e2de0504bad8b9253d5d/arch/arm64/core/Kconfig#L103-L104
https://github.com/zephyrproject-rtos/zephyr/blob/9c5abb35b5ffb0326e89e2de0504bad8b9253d5d/arch/x86/core/Kconfig.intel64#L20-L21
Simply special casing the architectures doesn't help, since (for example) an application that wants to push a UDP packet over a network may have different size dependencies depending on the underlying networking driver.

Thirdly, this is the same situation as the `HEAP_MEM_POOL_ADD_SIZE_`, a feature shared among an arbitrary number of users (not restricted to in-tree), where each user wants to impose some condition. For the heap the condition is that you want N bytes reserved for yourself. For the workqueue it is that the stack needs to be at least N bytes large. The only difference is the operation, MAX(n) vs SUM(n), so to me it makes sense to use the same type of solution.

Finally, only adding dependencies means that every application that uses a subsystem that requires a larger stack will need a `CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE` define (and possibly several compilation loops to figure out what that value should be). Once that process is done, you still don't know if that value will be always correct (see point two).

nordic-krch commented:
Can't we extend Kconfig or the standard is fixed? Ideally if we had something like:
```
imply SYSTEM_WORKQUEUE_STACK_SIZE >= 4096
```
that would be generic. I would also like to see something like `add FOO n`

JordanYates commented:
> Can't we extend Kconfig or the standard is fixed? Ideally if we had something like:
> 
> ```
> imply SYSTEM_WORKQUEUE_STACK_SIZE >= 4096
> ```
> 
> that would be generic. I would also like to see something like `add FOO n`

Extending is fine (e.g. #51316), changing the meaning of an existing keyword probably isn't.
The standard (last time I did it anyway) is that the Linux Kconfig tests still need to pass.

yperess commented:
> > Can't we extend Kconfig or the standard is fixed? Ideally if we had something like:
> > ```
> > imply SYSTEM_WORKQUEUE_STACK_SIZE >= 4096
> > ```
> > 
> > that would be generic. I would also like to see something like `add FOO n`
> 
> Extending is fine (e.g. #51316), changing the meaning of an existing keyword probably isn't. The standard (last time I did it anyway) is that the Linux Kconfig tests still need to pass.

I recognize that we've done it for heaps and I wish I had reviewed that when it landed. But that doesn't make this pattern a good one. Generally speaking magic names are a bad approach and I would much prefer using dependencies and forcing the application author to explicitly choose a stack size over accidentally getting a bigger size without knowing it.

Here's an example, say the app adds a feature which defines a magic config like this and adds 8kB of stack space. Suddenly your app drastically larger and you don't know exactly why. Further, lets say you added a feature and the magic config accidentally had a typo and adds too much stack space. Again, your app just gets bigger.

In my book I would prefer to get a build error saying a dependency failed and see the kconfig error saying the stack size must be at least `x`.

krish2718 commented on kernel/Kconfig:
```diff
@@ -577,10 +577,20 @@ rsource "Kconfig.obj_core"
-	default 2560 if WIFI_NM_WPA_SUPPLICANT
 	default 1024
 
+config SYSTEM_WORKQUEUE_STACK_SIZE_EXPLICIT
```
like heap, may be we can use same terminology?
```suggestion
config SYSTEM_WORKQUEUE_STACK_SIZE_IGNORE_MAX
```

JordanYates committed:
doc: document system workqueue stack size calculation

Make note of the new method of determining the system workqueue size in
the migration guide and the workqueue docs.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Docs added, `SYSTEM_WORKQUEUE_STACK_SIZE_EXPLICIT` renamed to `SYSTEM_WORKQUEUE_STACK_SIZE_IGNORE_MIN`.

pdgendt commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -26,6 +26,15 @@ Build System
+  among users as the maximum value of any Kconfig options whose name starts
+  with ``CONFIG_SYSTEM_WORKQUEUE_MIN_STACK_SIZE_``. The existing
+  :kconfig:option:`CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE` symbol value is copied
+  to :kconfig:option:`CONFIG_SYSTEM_WORKQUEUE_MIN_STACK_SIZE_DEFAULT`so that it
```
```suggestion
  to :kconfig:option:`CONFIG_SYSTEM_WORKQUEUE_MIN_STACK_SIZE_DEFAULT` so that it
```

pdgendt commented:
Maybe this needs a broader discussion, there are similar max/addition symbols that could benefit from an agreed mechanism. Like  `CONFIG_ZVFS_POLL_MAX` and probably others.

JordanYates commented:
> Maybe this needs a broader discussion, there are similar max/addition symbols that could benefit from an agreed upon mechanism. Like `CONFIG_ZVFS_POLL_MAX` and probably others.

Thats fair, if this was my only focus I would probably implement it as one or two additional Kconfig types that result in the desired value directly. e.g.:
```
config SYSTEM_WORKQUEUE_STACK_SIZE
  max "System workqueue stack size"
  default 1024 if ARCH_ARM
  default 2048 if ARCH_POSIX
  help
    Final value of the symbol is the maximum value of any of the `default` values with passing
    conditionals, or can be explicitly set to a value by assigning the value in prj.conf.

config HEAP_SIZE
  sum "System workqueue stack size"
  default 1024 if NETWORKING
  default 2048 if WPA_SUPPLICANT
  help
    Final value of the symbol is the total sum of any of the `default` values with passing conditionals,
    or can be explicitly set to a value by assigning the value in prj.conf
```
Unfortunately that is a larger project than I have time for at the moment.

cfriedt commented:
> Maybe this needs a broader discussion, there are similar max/addition symbols that could benefit from an agreed upon mechanism. Like `CONFIG_ZVFS_POLL_MAX` and probably others.

That actually has an implicit upper limit of `CONFIG_ZVFS_OPEN_MAX` too. It's a bit of a weird one.
